arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,310 @@
+vital::config_block_sptr
+aligned_edge_detection
+::get_configuration() const
+{
+  // Get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value( "threshold",
+                     d->threshold,
+                     "Minimum edge magnitude required to report as an edge "
+                     "in any output image." );
+  config->set_value( "produce_joint_output",
+                     d->produce_joint_output,
+                     "Set to false if we do not want to spend time computing "
+                     "joint edge images comprised of both horizontal and "
+                     "vertical information." );
+  config->set_value( "smoothing_sigma",
+                     d->smoothing_sigma,
+                     "Smoothing sigma for the output NMS edge density map." );
+  config->set_value( "smoothing_half_step",
+                     d->smoothing_half_step,
+                     "Smoothing half step for the output NMS edge density "
+                     "map." );
+
+  return config;
+}
+
+// ----------------------------------------------------------------------------
+void
+aligned_edge_detection
+::set_configuration( vital::config_block_sptr in_config )
+{
+  // Start with our generated vital::config_block to ensure that assumed values
+  // are present. An alternative would be to check for key presence before
+  // performing a get_value() call.
+  vital::config_block_sptr config = this->get_configuration();
+  config->merge_config( in_config );
+
+  // Settings for edge detection
+  d->threshold = config->get_value< float >( "threshold" );
+  d->produce_joint_output =
+    config->get_value< bool >( "produce_joint_output" );
+  d->smoothing_sigma = config->get_value< double >( "smoothing_sigma" );
+  d->smoothing_half_step =
+    config->get_value< unsigned >( "smoothing_half_step" );
+}
+
+// ----------------------------------------------------------------------------
+bool
+aligned_edge_detection
+::check_configuration( vital::config_block_sptr config ) const
+{
+  return true;
+}
+
+// ----------------------------------------------------------------------------
+kwiver::vital::image_container_sptr
+aligned_edge_detection
+::filter( kwiver::vital::image_container_sptr source_image_ptr )
+{
+  if( !source_image_ptr )
+  {
+    LOG_ERROR( logger(), "Image pointer was not valid." );
+    return nullptr;
+  }
+
+  // Get input image
+  vil_image_view_base_sptr source_image =
+    vxl::image_container::vital_to_vxl( source_image_ptr->get_image() );
+
+  // Perform Basic Validation
+  if( !source_image )
+  {
+    LOG_ERROR( logger(), "Image was not valid." );
+    return nullptr;
+  }
+
+  // Perform Basic Validation
+  if( source_image->nplanes() == 3 )
+  {
+    source_image = vil_convert_to_grey_using_average( source_image );
+  }
+  else if( source_image->nplanes() != 1 )
+  {
+    LOG_ERROR( logger(), "Input must be a grayscale image!" );
```

Comment:
Better error message might be `"Input must have either 1 or 3 channels but has " << source_image->nplanes()`